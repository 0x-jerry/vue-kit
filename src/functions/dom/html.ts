import { createAutoIncrementGenerator, Fn, isFn, isIterable, isObject } from '@0x-jerry/utils'
import { isElement } from '../utils'

/**
 * create a dom element
 *
 * @example
 *
 * ```ts
 * const click = () => console.log('cool')
 *
 * const el = html`
 * <div>
 * <button onclick=${click}> click me </button>
 * </div>
 * `
 *
 * document.body.append(el)
 * ```
 *
 * @param strings
 * @param values
 * @returns
 */
export function html(strings: TemplateStringsArray, ...values: any[]): DocumentFragment {
  const ctx = composeHtmlString(strings, ...values)

  const root = document.createElement('div')

  root.innerHTML = ctx.html

  walk(root, (node) => {
    if (node.nodeType === document.COMMENT_NODE) {
      const el = ctx.elements[node.textContent!.trim()]

      if (el) {
        node.parentElement?.replaceChild(el, node)
      }

      return
    }

    if (!isElement(node)) {
      return
    }

    for (const item of Array.from(node.attributes)) {
      let fn

      if (item.name.startsWith('on') && (fn = ctx.events[item.value])) {
        const eventName = item.name.slice(2)

        node.addEventListener(eventName, fn)
        node.removeAttribute(item.name)
      }
    }
  })

  const fragment = document.createDocumentFragment()

  fragment.append(...root.childNodes)

  return fragment
}

function walk(root: Node, cb: (node: Node) => void) {
  cb(root)

  for (const child of root.childNodes) {
    walk(child, cb)
  }
}

export function composeHtmlString(strings: TemplateStringsArray, ...values: any[]) {
  const nextId = createAutoIncrementGenerator('e')

  let _html = ''

  const events: Record<string, Fn> = {}

  const elements: Record<string, Node> = {}

  for (let idx = 0; idx < strings.length; idx++) {
    const str = strings[idx]
    const value = values[idx]

    _html += str

    if (idx < strings.length - 1) {
      concatValue(value)
    }
  }

  return {
    html: _html,
    events,
    elements,
  }

  function concatValue(value: any) {
    if (isObject(value) && isIterable(value)) {
      for (const item of value) {
        if (isFn(item)) {
          continue
        }

        concatValue(item)
      }

      return
    }

    if (value instanceof Node) {
      const id = nextId()
      _html += `<!-- ${id} -->`
      elements[id] = value
      return
    }

    if (isFn(value)) {
      const id = nextId()

      _html += `"${id}"`
      events[id] = value
      return
    }

    _html += String(value)
  }
}
