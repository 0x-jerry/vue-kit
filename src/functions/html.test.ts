// @jest-environment happy-dom

import { composeHtmlString, html } from './html'

describe('html tagged template', () => {
  it('should compose html', () => {
    const fn = vi.fn()
    const result = composeHtmlString`<div onclick=${fn}> hello ${2} </div>`

    expect(result).matchSnapshot()

    const result2 = composeHtmlString`<div onclick=${fn} dbclick=${fn} > hello ${2} </div>`

    expect(result2).matchSnapshot()
  })

  it('should bind click event', () => {
    const fn = vi.fn()

    const result = html`<div onclick=${fn} ondblclick=${fn}>hello ${2}</div>`

    expect(result).matchSnapshot()

    result.querySelector('div')?.dispatchEvent(new Event('click'))
    expect(fn).toBeCalledTimes(1)

    result.querySelector('div')?.dispatchEvent(new Event('dblclick'))
    expect(fn).toBeCalledTimes(2)
  })

  it('should respect HTMLElement', () => {
    const fn = vi.fn()

    const btn = html`<button onclick=${fn}>b1</button>`

    const result = html`<div>${btn}</div>`

    expect(result).matchSnapshot()

    result.querySelector('button')?.dispatchEvent(new Event('click'))
    expect(fn).toBeCalledTimes(1)
  })

  it('should respect Array', () => {
    const fn = vi.fn()

    const btns = [
      //
      html`<button onclick=${fn}>b1</button>`,
      html`<button onclick=${fn}>b2</button>`,
    ]

    const result = html`<div>
      ${btns}
      <!-- test -->
    </div>`

    expect(result).matchSnapshot()

    result.querySelectorAll('button').forEach((item) => {
      item.dispatchEvent(new Event('click'))
    })

    expect(fn).toBeCalledTimes(2)
  })
})
