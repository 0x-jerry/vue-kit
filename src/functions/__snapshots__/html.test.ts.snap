// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`html tagged template > should bind click event 1`] = `
<DocumentFragment>
  <div>
    hello 2
  </div>
</DocumentFragment>
`;

exports[`html tagged template > should compose html 1`] = `
{
  "elements": {},
  "events": {
    "e0": [MockFunction spy],
  },
  "html": "<div onclick=\\"e0\\"> hello 2 </div>",
}
`;

exports[`html tagged template > should compose html 2`] = `
{
  "elements": {},
  "events": {
    "e0": [MockFunction spy],
    "e1": [MockFunction spy],
  },
  "html": "<div onclick=\\"e0\\" dbclick=\\"e1\\" > hello 2 </div>",
}
`;

exports[`html tagged template > should respect Array 1`] = `
<DocumentFragment>
  <div>
    
      
    <button>
      b1
    </button>
    <button>
      b2
    </button>
    
      
    <!-- test -->
    
    
  </div>
</DocumentFragment>
`;

exports[`html tagged template > should respect HTMLElement 1`] = `
<DocumentFragment>
  <div>
    <button>
      b1
    </button>
  </div>
</DocumentFragment>
`;
